File: 2039-the-time-when-the-network-becomes-idle/2039-the-time-when-the-network-becomes-idle.cpp
class Solution {
public:
    int networkBecomesIdle(vector<vector<int>>& edges, vector<int>& patience) {
        int n = patience.size();
        vector <vector <int>> graph(n);
        vector <int> time(n, -1);
        
        for(auto x: edges) { // create adjacency list
            graph[x[0]].push_back(x[1]);
            graph[x[1]].push_back(x[0]);
        }
        int res = 0;
        queue <int> q;
        q.push(0);
        time[0] = 0;
        while(q.size()) {
            int node = q.front();
            q.pop();
            
            for(auto child: graph[node]) {
                if(time[child] == -1) { // if not visited.
                    time[child] = time[node] + 1; // calc time for child node
                    q.push(child);
					int d=time[child];
					int temp = 2*d;
             if(2*d > patience[child])
			 {
                if((2*d)%patience[child]==0)
                temp=4*d-patience[child];
                else
                    temp=4*d-(2*d)%patience[child];   
            }
            
            res = max(res,temp);
                }
            }
        }
        
        return res+1; // since till res time all messages has beed arrived so +1
    }
};

File: 2039-the-time-when-the-network-becomes-idle/2039-the-time-when-the-network-becomes-idle.java
class Solution {
    public int networkBecomesIdle(int[][] edges, int[] patience) {
        List<Integer>[] graph=new ArrayList[patience.length];
        for(int i=0;i<graph.length;i++) graph[i]=new ArrayList<>();
        for(int i=0;i<edges.length;i++) {
            graph[edges[i][0]].add(edges[i][1]);
            graph[edges[i][1]].add(edges[i][0]);
        }
        int[] dis=new int[patience.length];
        Arrays.fill(dis, -1);
        Queue<Integer> q=new LinkedList<>();
        for(int i=0;i<graph[0].size();i++) q.offer(graph[0].get(i));
        int len=1;
        while(!q.isEmpty()) {
            int size=q.size();
            for(int i=0;i<size;i++) {
                int cur=q.poll();
                if(dis[cur]!=-1) continue;
                dis[cur]=len;
                for(int j=0;j<graph[cur].size();j++) {
                    if(dis[graph[cur].get(j)]==-1) q.offer(graph[cur].get(j));
                }
            }
            len++;
        }
        int res=0;
        for(int i=1;i<dis.length;i++) {
            res=Math.max(res, patience[i]*((dis[i]*2-1)/patience[i])+dis[i]*2+1);
        }
        return res;
    }
}